concordium-node/src/grpc2.rs

```diff
@@ -0,0 +1,526 @@
+/// requires that we have all the data modelled in Rust. Our case is that the
+/// data that is encoded most of the time lives on the consensus/Haskell side of
+/// the FFI boundary. Thus it is natural to encode it there and only pass the
+/// byte array through FFI.. This presents a challenge since the grpc server is
+/// ultimately in Rust for various reasons. We solve the issue by using
+/// [tonic]'s support for custom [codecs](https://docs.rs/tonic/latest/tonic/codec/index.html).
+///
+/// In particular we define this "RawCodec" which knows how to encode messages
+/// that are already encoded, i.e., messages that are `AsRef<[u8]>`.
+///
+/// **NB: This codec does not know how to decode anything, and any attempt to do
+/// so will lead to an RPC error.**
+pub struct RawCodec<E, D>(PhantomData<E>, PhantomData<D>);
+impl<E, D> Default for RawCodec<E, D> {
+    fn default() -> Self { Self(Default::default(), Default::default()) }
+}
+
+impl<E: AsRef<[u8]> + Send + Sync + 'static, D: prost::Message + Default + 'static>
+    tonic::codec::Codec for RawCodec<E, D>
+{
+    type Decode = D;
+    type Decoder = RawDecoder<D>;
+    type Encode = E;
+    type Encoder = RawEncoder<E>;
+
+    fn encoder(&mut self) -> Self::Encoder { RawEncoder::default() }
+
+    fn decoder(&mut self) -> Self::Decoder { RawDecoder::default() }
+}
+
+pub struct RawEncoder<E>(PhantomData<E>);
+
+impl<E> Default for RawEncoder<E> {
+    fn default() -> Self { Self(Default::default()) }
+}
+
+impl<E: AsRef<[u8]>> tonic::codec::Encoder for RawEncoder<E> {
+    type Error = tonic::Status;
+    type Item = E;
+
+    fn encode(
+        &mut self,
+        item: Self::Item,
+        dst: &mut tonic::codec::EncodeBuf<'_>,
+    ) -> Result<(), Self::Error> {
+        dst.reserve(item.as_ref().len());
+        dst.put_slice(item.as_ref());
+        Ok(())
+    }
+}
+
+pub struct RawDecoder<D>(PhantomData<D>);
+
+impl<D> Default for RawDecoder<D> {
+    fn default() -> Self { RawDecoder(Default::default()) }
+}
+
+impl<D: prost::Message + Default> tonic::codec::Decoder for RawDecoder<D> {
+    type Error = tonic::Status;
+    type Item = D;
+
+    fn decode(
+        &mut self,
+        src: &mut tonic::codec::DecodeBuf<'_>,
+    ) -> Result<Option<Self::Item>, Self::Error> {
+        D::decode(src).map(Option::Some).map_err(|e| tonic::Status::internal(e.to_string()))
+    }
+}
+
+/// The implementation of the GRPC2 server.
+pub mod server {
+    use crate::{
+        configuration::GRPC2Config,
+        consensus_ffi::{consensus::ConsensusContainer, ffi::NotificationHandlers},
+        p2p::P2PNode,
+    };
+    use anyhow::Context;
+    use futures::{FutureExt, StreamExt};
+    use std::sync::{Arc, Mutex};
+    use tonic::{async_trait, transport::ServerTlsConfig};
+
+    use super::*;
+
+    /// An updatable list of listeners for events generated by the node.
+    /// The events which are naturally generated by the node
```

Comment:
What does it mean for the events to be naturally generated?

Reply:
I have elaborated.